userspace/async/async_metadata_source.h

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright (C) 2018 Sysdig, Inc.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <list>
+#include <map>
+#include <mutex>
+#include <thread>
+#include <stdint.h>
+
+namespace sysdig
+{
+
+/**
+ * Base class for classes that need to collect metadata asynchronously from some
+ * metadata source.  Subclasses will override the the run_impl() method.  In
+ * that method, subclasses will use use dequeue_next_key() method to get the
+ * key that it will use to collect the metadata, collect the appropriate
+ * metadata, and call the store_metadata() method to save the metadata.  The
+ * run_impl() method should continue to dequeue and process metadata while the
+ * queue_size() method returns non-zero.
+ *
+ * The constructor for this class accepts a maximum wait time; this specifies
+ * how long client code is willing to wait for a synchronous response (i.e.,
+ * how long the lookup() method will block waiting for the requested metadata).
+ * If the async_metadata_source is able to collect the requested metadata within
+ * that time period, then the lookup() method will return them.
+ *
+ * If the lookup() method is unable to collect the requested metadata within
+ * the requested time period, then one of two things will happen.  (1) If
+ * the client supplied a handler in the call to lookup(), then that handler
+ * will be invoked by the async_metadata_source once the metadata has been
+ * collected.  Note that the callback handler will be invoked in the context
+ * of the asynchronous thread associated with the async_metadata_source.  (2) If
+ * the client did not supply a handler, then the metadata will be stored, and the
+ * next call to the lookup() method with the same key will return the previously
+ * collected metadata.
+ *
+ * @tparam key_type      The type of the keys for which concrete subclasses will
+ *                       query.
+ * @tparam metadata_type The type of metadata that concrete subclasses will
+ *                       receive from a query.
+ */
+template<typename key_type, typename metadata_type>
+class async_metadata_source
```

Comment:
I'd rather this have its own semantics, which might be similar to map semantics, than try to overlay map semantics on this.

That said, I think renaming this is a reasonable idea.  I'll go with `async_key_value_source`.  I'm intentionally not calling it "store" since it's more than a "store".